README
NAME
    WWW::DanDomain - class to assist in interacting with DanDomain admin
    interface

VERSION
    This documentation describes version 0.07

SYNOPSIS
    The module lets the user interact with DanDomains administrative web
    interface. This can be used for automating tasks of processing data
    exports etc.

        use WWW::DanDomain;

        #All mandatory parameters
        #Please note DanDomain can be configured to use authorization on IP
        #meaning authentication is unnessesary
        my $wd = WWW::DanDomain->new({
            url => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        });

        #With optional authentication credentials
        my $wd = WWW::DanDomain->new({
            username => 'topshop',
            password => 'topsecret',
            url => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        });

        #with verbosity enabled
        my $wd = WWW::DanDomain->new({
            username => 'topshop',
            password => 'topsecret',
            url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
            verbose  => 1,
        });

        #With caching
        my $wd = WWW::DanDomain->new({
            username => 'topshop',
            password => 'topsecret',
            url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
            cache    => 1,
        });


        #With custom WWW::Mechanize object
        use WWW::Mechanize;

        my $mech = WWW::Mechanize->new(agent => 'MEGAnice bot');

        my $wd = WWW::DanDomain->new({
            username => 'topshop',
            password => 'topsecret',
            url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
            mech     => $mech,
        });
    
    
        #The intended use
        package My::WWW::DanDomain::Subclass;
    
        sub processor {
            my ( $self, $content ) = @_;
        
            #Note the lines terminations are Windows CRLF
            my @lines = split /\r\n/, $$content;
        
            ...
        
            }
        }
    
    
        #Using your new class
        my $my = My::WWW::DanDomain::Subclass->new({
            username => 'topshop',
            password => 'topsecret',
            url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        });
    
        my $content = $my->retrieve();
    
        print $$content;


        #Using a processor implemented as a code reference
        $wd = WWW::DanDomain->new({
            username  => 'topshop',
            password  => 'topsecret',
            url       => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
            processor => sub {                
                ${$_[0]} =~ s/test/fest/;        
                return $_[0];
            },
        });    


        #Implementing a processor class
        my $processor = MY::Processor->new();
    
        UNIVERSAL::can($processor, 'process');
    
        $wd = WWW::DanDomain->new({
            username  => 'topshop',
            password  => 'topsecret',
            url       => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
            processor => $processor,
        });
    
        my $content = $wd->retrieve();
    
        print ${$content};

DESCRIPTION
    This module is a simple wrapper around WWW::Mechanize it assists the
    user in getting going with automating tasks related to the DanDomain
    administrative web interface.

    Such as:

    *   manipulating data exports (removing, adjusting, calculating, adding
        columns)

    *   filling in missing data (combining data)

    *   converting formats (from CSV to XML, JSON, CSV, whatever)

METHODS
  new
    This is the constructor.

    The constructor takes a hash reference as input. The hash reference
    should contain keys according to the following conventions:

    *   username, optional username credential to access DanDomain

    *   password, optional password credential to access DanDomain

    *   url, the mandatory URL to retrieve data from ("retrieve")

    *   mech, a WWW::Mechanize object if you have a pre instantiated object
        or some other object implementing the the same API as
        WWW::Mechanize.

        The parameter is optional.

        See also cache parameter below for an example.

    *   verbose, a flag for indicating verbosity, default is 0 (disabled),
        the parameter is optional

    *   cache, usage of a cache meaning that we internally use
        WWW::Mechanize::Cached instead of WWW::Mechanize.

        The parameter is optional

    *   processor

        This parameter can be used of you do not want to implement a
        subclass of WWW::DanDomain.

        The processor parameter can either be:

        *   an object implementing a "proces" method, with the following
            profile:

                proces(\$content);

        *   a code reference with the same profile, adhering to the
            following example:

                sub { return ${$_[0]} };

  retrieve
    Parameters:

    *   a hash reference, the reference can be populated with statistic
        information based on the lineprocessing ("processor") initiated from
        "retrieve".

    The method returns a scalar reference to a string containing the content
    retrieved from the URL provided to the contructor ("new"). If the
    "processor" method is overwritten you can manipulate the content prior
    to being returned.

  process
    Takes the content retrieved (see: "retrieve") from the URL parameter
    provided to the constructor (see: "new"). You can overwrite the
    behaviour via the constructor (see: "new").

    Parameters:

    *   a scalar reference to a string to be processed line by line

    The stub does however not do anything, but it returns the scalar
    reference *untouched*.

  processor
    This is a wrapper for "process", provided for backwards compatibility.

DIAGNOSTICS
    *   Unable to authenticate, username and password not valid credentials

    *   Both username and password is required for authentication

        If you want to use authentication you have to provide both username
        and password.

    *   Unable to retrieve base URL: $@

        The base URL provided to retrieve gives an error.

        Please see: <http://search.cpan.org/perldoc?HTTP%3A%3AResponse> or
        <http://search.cpan.org/~gaas/libwww-perl/lib/HTTP/Status.pm>

        Test the URL in your browser to investigate.

    *   Unable to retrieve URL: $@

        The base URL provided to retrieve gives an error.

        Please see: <http://search.cpan.org/perldoc?HTTP%3A%3AResponse> or
        <http://search.cpan.org/~gaas/libwww-perl/lib/HTTP/Status.pm>

        Test the URL in your browser to investigate.

CONFIGURATION AND ENVIRONMENT
    The module requires Internet access to make sense and an account with
    DanDomain with username and password is required.

DEPENDENCIES
    *   WWW::Mechanize

    *   Carp

TEST AND QUALITY
    The tests are based on Test::MockObject::Extends and example data are
    mocked dummy data. Please see the TODO section.

    The test suite uses the following environment variables as flags:

    TEST_AUTHOR, to test prerequisites, using Test::Prereq
    TEST_CRITIC, to do a static analysis of the code, using Perl::Critic,
    see also QUALITY AND CODING STANDARD

  TEST COVERAGE
    The following data are based on an analysis created using Devel::Cover
    and the distributions own test suite, instantiated the following way.

        % ./Build testcover --verbose

        ---------------------------- ------ ------ ------ ------ ------ ------ ------
        File                           stmt   bran   cond    sub    pod   time  total
        ---------------------------- ------ ------ ------ ------ ------ ------ ------
        blib/lib/WWW/DanDomain.pm     100.0  100.0  100.0  100.0  100.0  100.0  100.0
        Total                         100.0  100.0  100.0  100.0  100.0  100.0  100.0
        ---------------------------- ------ ------ ------ ------ ------ ------ ------

    Please note the report is based on version 0.03 of WWW::DanDomain

QUALITY AND CODING STANDARD
    The code passes Perl::Critic tests a severity: 1 (brutal)

    The following policies have been disabled:

    Perl::Critic::Policy::InputOutput::RequireBracedFileHandleWithPrint

    Perl::Critic resource file, can be located in the t/ directory of the
    distribution t/perlcriticrc

    Perl::Tidy resource file, can be obtained from the original author

BUGS AND LIMITATIONS
    No known bugs at this time.

BUG REPORTING
    Please report any bugs or feature requests via:

    *   email: "bug-www-dandomain at rt.cpan.org"

    *   HTTP: <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-DanDomain>

DEVELOPMENT
    *   Subversion repository: <http://logicLAB.jira.com/svn/DAND>

TODO
    *   Most of the work is done in the classes inheriting from this class,
        there could however be work to do in the maintenance area, making
        this class more informative if failing

    *   I would like to add some integration test scripts so I can see that
        the package works with real data apart from the mock.

SEE ALSO
    *   <http://www.dandomain.dk>

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc WWW::DanDomain

    You can also look for information at:

    *   Official Wiki

        <http://logiclab.jira.com/wiki/display/DAND/Home+-+WWW-DanDomain>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/WWW-DanDomain>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/WWW-DanDomain>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-DanDomain>

    *   Search CPAN

        <http://search.cpan.org/dist/WWW-DanDomain>

AUTHOR
    *   jonasbn, "<jonasbn at cpan.org>"

MOTIVATION
    This module grew out of a small script using WWW::Mechanize to fetch
    some data from a website and changing it to satisfy the client utilizing
    the data.

    More a more scripts where based on the original script giving a lot of
    redundant code. Finally I refactored the lot to use some common code
    base.

    After some time I refactored to an object oriented structure making it
    even easier to maintain and adding more clients. This made the actual
    connectivity into a package (this package) letting it loose as open
    source.

ACKNOWLEDGEMENTS
    *   Andy Lester (petdance) the author of WWW::Mechanize and
        WWW::Mechanize:Cached, this module makes easy things easy and hard
        things possible.

    *   Steen Schnack, who understand the power and flexibility of computer
        programming and custom solutions and who gave me the assignment.

    *   Slaven Rezic, for his CPAN testers work and assistance in pointing
        out an issue with release 0.03

LICENSE AND COPYRIGHT
    Copyright 2009-2014 jonasbn, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Change history for WWW::DanDomain

Project change log:
http://logicLAB.jira.com/browse/DAND#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel

0.07 2014-08-23, maintenance release, updated not required

- Added specification of Perl requirement as 5.8 as pointed out
  by Perl::Critic

- Added changes.t automatic assertion of the integrity of the
  Changes file see also DAND-10 [DAND-11]

  https://logiclab.jira.com/browse/MBB-40


0.06 2013-07-24, maintenance release, updated not required

- Fixed up Changes file as part of my Questhub quest adhering to
  the standard described in: CPAN::Changes::Spec [DAND-10]

  Quest:
  http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec

  https://logiclab.jira.com/browse/DAND-10


0.05 2011-04-17, Feature release, updated not required

- Introducing new parameter to constructor, providing more
  flexibility in interacting with the class

  http://logicLAB.jira.com/browse/DAND-8 [DAND-8]


0.04 2010-02-21, Bug fix release, updated not required

- Addressing failing test report

    WWW-DanDomain-0.03:
    - i386-freebsd / 5.10.1:
    - FAIL http://www.cpantesters.org/cpan/report/6848267


0.03 2010-02-20, Maintenance release, updated not required

- Implemented handling of DanDomain access scenario where
  authentication credentials are not needed, since authorization
  is based on IP validation (feature with DanDomain)

- Improved error handling


0.02 2009-09-16, Minor release feature/bug fix release, update not required

- Fixed dependencies, WWW::Mechanize was not listed, so smoke
  tests were failing

- First shot at caching using WWW::Mechanize::Cached


0.01 2009-06-15

- First version, to be released on an unsuspecting world.


Build.PL
# Build.PL for Spilpriser

# $Id: Build.PL 383 2009-01-08 12:39:01Z jonasbn $

use strict;
use Module::Build;

my $build = Module::Build->new(
    dist_author          => 'Jonas B. Nielsen, <jonasbn@logicLAB.dk>',
    module_name          => 'WWW::DanDomain',
    dist_version_from    => 'lib/WWW/DanDomain.pm',
    dist_abstract        => 'WWW::DanDomain',
    license              => 'perl',
    recursive_test_files => 1,
    create_readme        => 1,
    configure_requires   => {},
    create_makefile_pl   => 'traditional',
    build_requires       => {
        'Test::More'                => '0', 
        'File::Slurp'               => '0',
        'Test::Exception'           => '0',
        'Test::MockObject::Extends' => '0',
        'Test::Perl::Critic'        => '0',
        'Test::Prereq::Build'       => '0',
        'Env'                       => '0',
    },
    requires => {
        'Module::Build'             => '0.30',
        'Data::Dumper'           => '0',    #Debugging
        'WWW::Mechanize'         => '0',
        'WWW::Mechanize::Cached' => '0',
        'Carp'                   => '0',
        'Try::Tiny'              => '0',
        'UNIVERSAL::can'         => '0',
        'perl'                   => '5.008', 
    },
    meta_merge => {
        resources => {
            homepage => 'http://logicLAB.jira.com/browse/DAND',
            bugtracker =>
                'http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-DanDomain',
            repository => 'https://github.com/jonasbn/dand',
            ProjectChangelog =>
                'http://logiclab.jira.com/browse/DAND#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel',
            ProjectSite => 'http://logicLAB.jira.com/browse/DAND',
            ProjectWiki => 'http://logiclab.jira.com/wiki/display/DAND/Home',
        }
    },

);

$build->create_build_script();

1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.008;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'WWW::DanDomain',
  'VERSION_FROM' => 'lib/WWW/DanDomain.pm',
  'PREREQ_PM' => {
                   'Carp' => '0',
                   'Data::Dumper' => '0',
                   'Env' => 0,
                   'File::Slurp' => 0,
                   'Module::Build' => '0.30',
                   'Test::Exception' => 0,
                   'Test::MockObject::Extends' => 0,
                   'Test::More' => 0,
                   'Test::Perl::Critic' => 0,
                   'Test::Prereq::Build' => 0,
                   'Try::Tiny' => '0',
                   'UNIVERSAL::can' => '0',
                   'WWW::Mechanize' => '0',
                   'WWW::Mechanize::Cached' => '0'
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {},
  'test' => {
              'TESTS' => 't/*.t'
            }
)
;


lib/WWW/DanDomain.pm
package WWW::DanDomain;

use warnings;
use strict;
use WWW::Mechanize;
use WWW::Mechanize::Cached;
use Carp qw(croak);
use Try::Tiny;
use UNIVERSAL::can;
use 5.008;

our $VERSION = '0.07';

sub new {
    my ( $class, $param ) = @_;

    my $mech;

    my $agent = __PACKAGE__ . "-$VERSION";

    if ( $param->{mech} ) {
        $mech = $param->{mech};
    } elsif ( $param->{cache} ) {
        $mech = WWW::Mechanize::Cached->new( agent => $agent );
    } else {
        $mech = WWW::Mechanize->new( agent => $agent );
    }

    my $self = bless {
        base_url  => 'http://www.billigespil.dk/admin',
        username  => $param->{username},
        password  => $param->{password},
        url       => $param->{url},
        verbose   => $param->{verbose} || 0,
        mech      => $mech,
        processor => $param->{processor},
    }, $class;

    return $self;
}

sub retrieve {
    my ( $self, $stat ) = @_;

    if ( $self->{username} || $self->{password} ) {
        if ( ( not $self->{username} ) or ( not $self->{password} ) ) {
            croak 'Both username and password is required for authentication';
        }

        $self->{mech}->get( $self->{base_url} )
            or croak "Unable to retrieve base URL: $@";

        if (not $self->{mech}->submit_form(
                form_number => 0,
                fields      => {
                    UserName => $self->{username},
                    Password => $self->{password},
                }
            )
            )
        {
            croak 'Unable to authenticate';
        }
    }

    $self->{mech}->get( $self->{url} ) or croak "Unable to retrieve URL: $@";

    my $content = $self->{mech}->content();

    if ( ref $self->{processor} eq 'CODE' ) {
        return &{ $self->{processor} }( \$content, $stat );
    } elsif ( ref $self->{processor} ) {
        try {
            $self->{processor}->can('process');
        }
        catch {
            croak q{Your processor does not implement 'process' method};
        };
        return $self->{processor}->process( \$content, $stat );
    } else {
        return $self->process( \$content, $stat );
    }
}

sub processor {
    my ( $self, $content ) = @_;

    return $self->process($content);
}

sub process {
    my ( $self, $content ) = @_;

    if ( $self->{verbose} ) {
        print STDERR "${$content}.\n";
    }

    return $content;
}

1;

__END__

=pod

=head1 NAME

WWW::DanDomain - class to assist in interacting with DanDomain admin interface

=head1 VERSION

This documentation describes version 0.07

=head1 SYNOPSIS

The module lets the user interact with DanDomains administrative web interface.
This can be used for automating tasks of processing data exports etc.

    use WWW::DanDomain;

    #All mandatory parameters
    #Please note DanDomain can be configured to use authorization on IP
    #meaning authentication is unnessesary
    my $wd = WWW::DanDomain->new({
        url => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
    });

    #With optional authentication credentials
    my $wd = WWW::DanDomain->new({
        username => 'topshop',
        password => 'topsecret',
        url => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
    });

    #with verbosity enabled
    my $wd = WWW::DanDomain->new({
        username => 'topshop',
        password => 'topsecret',
        url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        verbose  => 1,
    });

    #With caching
    my $wd = WWW::DanDomain->new({
        username => 'topshop',
        password => 'topsecret',
        url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        cache    => 1,
    });


    #With custom WWW::Mechanize object
    use WWW::Mechanize;

    my $mech = WWW::Mechanize->new(agent => 'MEGAnice bot');

    my $wd = WWW::DanDomain->new({
        username => 'topshop',
        password => 'topsecret',
        url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        mech     => $mech,
    });
    
    
    #The intended use
    package My::WWW::DanDomain::Subclass;
    
    sub processor {
        my ( $self, $content ) = @_;
        
        #Note the lines terminations are Windows CRLF
        my @lines = split /\r\n/, $$content;
        
        ...
        
        }
    }
    
    
    #Using your new class
    my $my = My::WWW::DanDomain::Subclass->new({
        username => 'topshop',
        password => 'topsecret',
        url      => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
    });
    
    my $content = $my->retrieve();
    
    print $$content;


    #Using a processor implemented as a code reference
    $wd = WWW::DanDomain->new({
        username  => 'topshop',
        password  => 'topsecret',
        url       => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        processor => sub {                
            ${$_[0]} =~ s/test/fest/;        
            return $_[0];
        },
    });    


    #Implementing a processor class
    my $processor = MY::Processor->new();
    
    UNIVERSAL::can($processor, 'process');
    
    $wd = WWW::DanDomain->new({
        username  => 'topshop',
        password  => 'topsecret',
        url       => 'http://www.billigespil.dk/admin/edbpriser-export.asp',
        processor => $processor,
    });
    
    my $content = $wd->retrieve();
    
    print ${$content};

=head1 DESCRIPTION

This module is a simple wrapper around L<WWW::Mechanize> it assists the user
in getting going with automating tasks related to the DanDomain administrative
web interface.

Such as:

=over

=item * manipulating data exports (removing, adjusting, calculating, adding
columns)

=item * filling in missing data (combining data)

=item * converting formats (from CSV to XML, JSON, CSV, whatever)

=back

=head1 METHODS

=head2 new

This is the constructor.

The constructor takes a hash reference as input. The hash reference should
contain keys according to the following conventions:

=over

=item * username, optional username credential to access DanDomain 

=item * password, optional password credential to access DanDomain

=item * url, the mandatory URL to retrieve data from (L</retrieve>)

=item * mech, a L<WWW::Mechanize> object if you have a pre instantiated object
or some other object implementing the the same API as L<WWW::Mechanize>.

The parameter is optional. 

See also cache parameter below for an example.

=item * verbose, a flag for indicating verbosity, default is 0 (disabled), the
parameter is optional

=item * cache, usage of a cache meaning that we internally use
L<WWW::Mechanize::Cached> instead of L<WWW::Mechanize>.

The parameter is optional

=item * processor

This parameter can be used of you do not want to implement a subclass of
WWW::DanDomain.

The processor parameter can either be:

=over

=item * an object implementing a L</proces> method, with the following profile:

    proces(\$content);

=item * a code reference with the same profile, adhering to the following example:

    sub { return ${$_[0]} };

=back

=back

=head2 retrieve

Parameters:

=over

=item * a hash reference, the reference can be populated with statistic
information based on the lineprocessing (L</processor>) initiated from
L</retrieve>.

=back

The method returns a scalar reference to a string containing the content
retrieved from the URL provided to the contructor (L</new>). If the
L</processor> method is overwritten you can manipulate the content prior
to being returned.

=head2 process

Takes the content retrieved (see: L</retrieve>) from the URL parameter provided
to the constructor (see: L</new>). You can overwrite the behaviour via the
constructor (see: L</new>).

Parameters:

=over

=item * a scalar reference to a string to be processed line by line

=back

The stub does however not do anything, but it returns the scalar reference
I<untouched>.

=head2 processor

This is a wrapper for L</process>, provided for backwards compatibility.

=head1 DIAGNOSTICS

=over

=item * Unable to authenticate, username and password not valid credentials

=item * Both username and password is required for authentication

If you want to use authentication you have to provide both B<username> and
B<password>.

=item * Unable to retrieve base URL: $@

The base URL provided to retrieve gives an error.

Please see: L<http://search.cpan.org/perldoc?HTTP%3A%3AResponse> or
L<http://search.cpan.org/~gaas/libwww-perl/lib/HTTP/Status.pm>

Test the URL in your browser to investigate.

=item * Unable to retrieve URL: $@

The base URL provided to retrieve gives an error.

Please see: L<http://search.cpan.org/perldoc?HTTP%3A%3AResponse> or
L<http://search.cpan.org/~gaas/libwww-perl/lib/HTTP/Status.pm>

Test the URL in your browser to investigate.

=back

=head1 CONFIGURATION AND ENVIRONMENT

The module requires Internet access to make sense and an account with DanDomain
with username and password is required.

=head1 DEPENDENCIES

=over

=item * L<WWW::Mechanize>

=item * L<Carp>

=back

=head1 TEST AND QUALITY

The tests are based on L<Test::MockObject::Extends> and example data are
mocked dummy data. Please see the TODO section.

The test suite uses the following environment variables as flags:

=over

=item TEST_AUTHOR, to test prerequisites, using L<Test::Prereq>

=item TEST_CRITIC, to do a static analysis of the code, using L<Perl::Critic>,
see also QUALITY AND CODING STANDARD

=back

=head2 TEST COVERAGE

The following data are based on an analysis created using L<Devel::Cover> and
the distributions own test suite, instantiated the following way.

    % ./Build testcover --verbose

    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    File                           stmt   bran   cond    sub    pod   time  total
    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    blib/lib/WWW/DanDomain.pm     100.0  100.0  100.0  100.0  100.0  100.0  100.0
    Total                         100.0  100.0  100.0  100.0  100.0  100.0  100.0
    ---------------------------- ------ ------ ------ ------ ------ ------ ------

Please note the report is based on version 0.03 of WWW::DanDomain

=head1 QUALITY AND CODING STANDARD

The code passes L<Perl::Critic> tests a severity: 1 (brutal)

The following policies have been disabled:

=over

=item L<Perl::Critic::Policy::InputOutput::RequireBracedFileHandleWithPrint>

=back

L<Perl::Critic> resource file, can be located in the t/ directory of the
distribution F<t/perlcriticrc>

L<Perl::Tidy> resource file, can be obtained from the original author

=head1 BUGS AND LIMITATIONS

No known bugs at this time.

=head1 BUG REPORTING

Please report any bugs or feature requests via:

=over

=item * email: C<bug-www-dandomain at rt.cpan.org>

=item * HTTP: L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-DanDomain>

=back

=head1 DEVELOPMENT

=over

=item * Subversion repository: L<http://logicLAB.jira.com/svn/DAND>

=back

=head1 TODO

=over

=item * Most of the work is done in the classes inheriting from this class,
there could however be work to do in the maintenance area, making this class
more informative if failing

=item * I would like to add some integration test scripts so I can see that the
package works with real data apart from the mock.

=back

=head1 SEE ALSO

=over

=item * L<http://www.dandomain.dk>

=back

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::DanDomain

You can also look for information at:

=over 4

=item * Official Wiki

L<http://logiclab.jira.com/wiki/display/DAND/Home+-+WWW-DanDomain>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-DanDomain>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-DanDomain>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-DanDomain>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-DanDomain>

=back

=head1 AUTHOR

=over

=item * jonasbn, C<< <jonasbn at cpan.org> >>

=back

=head1 MOTIVATION

This module grew out of a small script using L<WWW::Mechanize> to fetch some
data from a website and changing it to satisfy the client utilizing the data.

More a more scripts where based on the original script giving a lot of redundant
code. Finally I refactored the lot to use some common code base.

After some time I refactored to an object oriented structure making it even
easier to maintain and adding more clients. This made the actual connectivity
into a package (this package) letting it loose as open source.

=head1 ACKNOWLEDGEMENTS

=over

=item * Andy Lester (petdance) the author of L<WWW::Mechanize> and
L<WWW::Mechanize:Cached>, this module makes easy things easy and hard things
possible.

=item * Steen Schnack, who understand the power and flexibility of computer
programming and custom solutions and who gave me the assignment.

=item * Slaven Rezic, for his CPAN testers work and assistance in pointing out
an issue with release 0.03

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2009-2014 jonasbn, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
